### PR TITLE
Roles - Initial start

### PR DESCRIPTION
**Pull request by JamesSCBD**

### CODE REVIEW

File: middleware/roles.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+export default ({ route, $auth, redirect }) => {
+  const options = getRouteComponentOptions(route, 'auth', 'roles');
+
+  if (options.auth === false || options.auth === 'guest') return;
+
+  const { roles } = options;
+
+  if (!roles) throw new Error('Role is not configured.');
+
+  if (roles === []) return;
```
Comment (sbilodeau) [inline middleware/roles.js:10]: this is wrong.
two arrays cannot be ===

```
[] === [] => false!!!
```

you have to test: isEmptyor lenght==0

File: pages/contacts/index.vue [hunk 2]
```diff
@@ -268,7 +268,7 @@ export default {
       isContactSelected: 'contacts/isContactSelected',
     }),
   },
-  auth   : readOnly,
+  roles  : [ ...Administrators, ...UsersCanRead ],
```
Comment (sbilodeau) [inline pages/contacts/index.vue:271]: onle is enough: UsersCanRead already contains admins'

File: pages/forbidden.vue [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    <h3>Sorry, you are not allowed to access this page.</h3>
+  </div>
+</template>
+
+<script>
+import { Anyone } from '@roles';
+
+export default {
+  name : 'Forbidden',
+  roles: [ ...Anyone ],
```
Comment (sbilodeau) [inline pages/forbidden.vue:12]: no. use `auth:false` instead

**Review by sbilodeau — CHANGES_REQUESTED** [review]
Please delete old set of roles
Implement left menu security
make sure you check security for batch tasks. I cannot see it

File: middleware/roles.js (hunk 1 above)
Comment (JamesSCBD) [reply middleware/roles.js:10]: yes true... my bad.

**Review by sbilodeau — CHANGES_REQUESTED** [review]

File: configs/roles.js [hunk 4]
```diff
@@ -6,3 +7,12 @@ export const meetingNominator = [ 'Kronos-FullAccess', 'Kronos-Meeting-Nominator
...
 export const admins = fullAccess;
+
+export const Anyone = [ 'Kronos-FullAccess', 'Kronos-User', 'Kronos-Meeting-Nominator', 'Kronos-ReadOnly' ];
+export const Administrators = [ 'Kronos-FullAccess' ];
+export const Users = [ 'Kronos-FullAccess', 'Kronos-User' ];
+export const UsersCanRead = [ 'Kronos-FullAccess', 'Kronos-User', 'Kronos-ReadOnly' ];
+export const UsersCanDelete = [ 'Kronos-FullAccess' ];
+export const UsersCanValidate = [ 'Kronos-FullAccess' ];
+export const UsersCanExport = [ 'Kronos-FullAccess' ];
+export const MeetingNominators = [ 'Kronos-FullAccess', 'Kronos-User', 'Kronos-Meeting-Nominator' ];
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply configs/roles.js]: DONE!

File: middleware/roles.js (hunk 1 above)
Comment (JamesSCBD) [reply middleware/roles.js]: DONE!

File: pages/contacts/index.vue (hunk 2 above)
Comment (JamesSCBD) [reply pages/contacts/index.vue]: DONE!

File: pages/forbidden.vue (hunk 3 above)
Comment (JamesSCBD) [reply pages/forbidden.vue]: DONE!

File: components/kronos/sidebar/Nav.vue [hunk 5]
```diff
@@ -1,12 +1,12 @@
 <template>
   <CSidebarNav>
     <BFormSelect id="page-size" :value="conferenceId" class="form-control" :options="conferenceOptions" @change="setSelectedConference" />
-    <CSidebarNavItem name="Dashbord" icon="tachometerAlt" :to="`/`" />
+    <CSidebarNavItem v-if="accessibleRoutes.includes('index')" name="Dashbord" icon="tachometerAlt" @click.native="redirect('index')" />
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply components/kronos/sidebar/Nav.vue]: DONE!

File: components/kronos/sidebar/Nav.vue [hunk 6]
```diff
@@ -16,26 +16,41 @@ import {
...
     }),
+    redirect(routeName) {
+      this.$router.push({ name: routeName });
+    },
   },
 };
 
+async function created() {
+  const pageRoutes      = [ 'index', 'contacts', 'organizations', 'conferences', 'reports' ];
+  this.accessibleRoutes = await getAccessibleRoutes.call(this, pageRoutes);
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply components/kronos/sidebar/Nav.vue]: DONE!

File: components/list/columns/MemberCount.vue [hunk 7]
```diff
@@ -14,8 +14,17 @@
...
-  methods   : {
+  props     : {
+    organization: {
+      type    : Object,
+      required: true,
+    },
+    addAccess: {
+      type   : Boolean,
+      default: true,
+    },
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply components/list/columns/MemberCount.vue]: DONE!

File: components/list/columns/MemberCount.vue [hunk 8]
```diff
@@ -1,9 +1,9 @@
 <template>
   <div class="row align-items-center">
-    <div class="col-6">
+    <div class="text-center" :class="addAccess ? 'col-6': 'col-12'">
       {{ organization && organization.memberCount }}
     </div>
-    <div class="col-4">
+    <div v-if="addAccess" class="col-4 text-center">
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply components/list/columns/MemberCount.vue]: DONE!

File: components/list/columns/MemberCount.vue [hunk 9]
```diff
@@ -1,9 +1,9 @@
 <template>
   <div class="row align-items-center">
-    <div class="col-6">
+    <div class="text-center" :class="addAccess ? 'col-6': 'col-12'">
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply components/list/columns/MemberCount.vue]: DONE!

File: middleware/roles.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
+export default ({ route, $auth, redirect }) => {
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply middleware/roles.js]: DONE!

File: pages/organizations/index.vue [hunk 11]
```diff
@@ -413,7 +418,11 @@ function cleanUp(obj) {
 }
 
 async function rowDoubleClicked(item) {
-  this.$router.push(`organizations/${item.organizationId}`);
+  const routeName = 'organizations-organizationId';
+  if (!this.accessibleRoutes.includes(routeName)) return;
+
+  const { organizationId } = item;
+  this.$router.push({ name: 'organizations-organizationId', params: { organizationId } });
```
[the remark this answers is not in the document]
Comment (JamesSCBD) [reply pages/organizations/index.vue]: DONE!

**Pull request merged**